blog/2024-09-02-amqp-flow-control/index.md

```diff
@@ -0,0 +1,762 @@
++------|------+                                  |  +----------------------------------+
+       |                                         |
+       |                                         |
+Publisher or Consumer                      AMQP 1.0 link
+   goroutine
+```
+
+Run the benchmark as follows:
+1. Start RabbitMQ server [v4.0.0-beta.6](https://github.com/rabbitmq/rabbitmq-server/releases/tag/v4.0.0-beta.6) using `make run-broker` on an Ubuntu box.
+2. Execute the Go program with `go run one_sender_one_receiver/main.go`
+3. Once the program completes, list the number of messages in each queue:
+```
+./sbin/rabbitmqctl --silent list_queues name type messages --formatter=pretty_table
+┌────────────────────────┬─────────┬──────────┐
+│ name                   │ type    │ messages │
+├────────────────────────┼─────────┼──────────┤
+│ classic-queue-amqp-091 │ classic │ 990932   │
+├────────────────────────┼─────────┼──────────┤
+│ quorum-queue-amqp-091  │ quorum  │ 172800   │
+├────────────────────────┼─────────┼──────────┤
+│ classic-queue-amqp-10  │ classic │ 336229   │
+├────────────────────────┼─────────┼──────────┤
+│ quorum-queue-amqp-10   │ quorum  │ 130000   │
+└────────────────────────┴─────────┴──────────┘
+```
+
+While the AMQP 0.9.1 client consumes only `1,000,000 - 990,932 = 9,068` messages, the AMQP 1.0 client consumes `1,000,000 - 336,229 = 663,771` messages.
+
+Therefore, in this benchmark, the AMQP 1.0 client receives 73 times (!) more messages than the AMQP 0.9.1 client.
+
+</details>
+
+:::
+
+:::note
+
+In AMQP 0.9.1, [consumer prefetch](/docs/consumer-prefetch) limits the number of unacknowledged messages.
+When a consumer acknowledges messages by sending `basic.ack` frames, RabbitMQ delivers additional messages.
+
+In AMQP 1.0, message acknowledgment is independent of link flow control.
+A consumer acknowledging messages by sending [disposition](https://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-transport-v1.0-os.html#type-disposition) frames will not prompt RabbitMQ to deliver more messages.
+Instead, the client must replenish link credit by sending `flow` frames for RabbitMQ to continue sending messages.
+For convenience, some AMQP 1.0 client libraries automatically send both `disposition` and `flow` frames when your application acknowledges messages.
+
+:::
+
+### Delivery-Count
+
+So far, we understand only one field of the [flow](https://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-transport-v1.0-os.html#type-flow) frame: `link-credit`.
+
+What happens in the following scenario?
+```
+Receiver                                  Sender
+=======================================================================
+                             ...
+link state:                               link state:
+ link-credit = 3                           link-credit = 3
+
+flow(link-credit = 6)     ---+   +---     transfer(...)
+                             \ /
+                              x
+                             / \
```

Comment:
The lines look misaligned with the horizontal parts.

Reply:
Good spot.